### ISSUE TITLE
Problem with states when an adapter is applied

### ISSUE THREAD

**Issue by GianlucaVesentini** [amcharts/amcharts4#3607]
Hi, 

I followed the tutorial ["Adaptive column corner rounding for positive and negative values"](https://www.amcharts.com/docs/v4/tutorials/adaptive-column-corner-rounding-for-positive-and-negative-values/) and I created a graph just like the one in the example.

Then I wanted to add some effects for the colums that are hovered by the cursor of the mouse, so I took a look at the page [States](https://www.amcharts.com/docs/v4/concepts/states/).

The property "fillOpacity", which is not affected by an adapter, is changing according to my code.
The properties "cornerRadiusTopLeft" and "cornerRadiusTopRight" instead are not updating, unless I remove the adapter on those properties.

The code is the following: 
https://codepen.io/GianlucaV/pen/zYZgjBW

I expected the "default" state to save the values of the properties applied by the adapters while we are in the hover state, but instead it seems that the adapters are overriding the values applied by states.

Is there a way to apply both Adapters and States?

Thank you,
Gianluca

**Comment by GianlucaVesentini** [comment]
Thanks fo the reply.

I updated the code and I am almost done, but there are some things that do not make sense to me.

Code (same link): https://codepen.io/GianlucaV/pen/zYZgjBW  

**First:** Adapter alone does not work, it seems that I still need a sort of trigger for the adapter function, thus I am forced to include:

- line 78: `var hoverState = series.columns.template.column.states.create("hover"); `

- line 81: `hoverState.properties.fillOpacity = 1;`

Without these lines it is pointless to check if the target is hovered in the fillOpacity adapter because it only triggers when the graph is created.

**Second:** I can't understand the behavior of the following line in the fillOpacity adapter function:

- line 105:  `return (target && target.isHover) ? 0.7: 1;`

When the graph is created, fillOpacity is set to 1 according to this line because targets are not hovered. When you hover a column, you can see that it flickers (I assume line 105 returns the value 0.7 for a moment, then line 81 takes over for some unexpected reasons or adapters triggers again but this time 1 is returned instead, i'm not able to tell).
When the column is no more hovered, adapters returns value 0.7, which contradicts the logic of the condition that I set in line 105.
From this point onwards the behaviour of the column is exactly the one that I want: fillOpacity is set to 0.7 when not hovered and 1 when hovered.

Now it should be clear that the behavior that I want is the opposite of what I stated in line 105, but the point is that if I replace line 105 with the correct one `return (target && target.isHover) ? 1: 0.7;` I obtain that:

1. fillOpacity property never changes and it is always set to 0.7
2. the first time that I hover a column only corner radiuses change but then they never return to their original value

To remove the initial flicker is sufficient, but there is the need to correct also the logic of the code.

**Third:** this is not important but if a set corner radiuses inside the adapter the animations are not smooth. I don't care about this one but i wanted to inform you.

I'm sorry if I am bringing a lot of problems, but I think that maybe others may face the same issues.

**Comment by zeroin** [comment]
> Adapter alone does not work, it seems that I still need a sort of trigger for the adapter function, thus I am forced to include:

Yes, this is correct behavior. Adapter is used only when a sprite is redrawn, and in case nothing changes, it is not used. So when you add modify some property on hover, you force the sprite to be redrawn, and your adapter works.

I made some modifications:
https://codepen.io/team/amcharts/pen/04a3a9188f39e73210998706312d1d18

I set initial `fillOpacity` to 1 and hover `fillOpacity` to 0.7. and then the `fillOpacity` adapter returns the same value which was passed to adapter, so we use it only for modifying radiuses. 

Animation only works between states, adapter sets value without animation.
